src/object_based/extent_htm.cpp

```diff
@@ -0,0 +1,352 @@
+    if (wf.is_circular) {
+      // simplified dot and norm of cross product assuming that circle_centre
+      // is {0, 1, 0}
+      double dot = position_t.y();
+
+      return {wf.angle_to_weight.inside_from_cos(dot),
+              wf.angle_to_weight.outside_from_cos(dot)};
+    }
+
+    Eigen::RowVector3d position_t_right = position_t;
+    position_t_right(0) = std::abs(position_t(0));
+
+    // for the straight lines
+    if (position_t_right.head<2>() * wf.circle_test >= 0) {
+      double sin_abs_elevation = std::abs(position_t(2));
+      return {wf.angle_to_weight.inside_from_sin(sin_abs_elevation),
+              wf.angle_to_weight.outside_from_sin(sin_abs_elevation)};
+    } else {
+      double dot = wf.right_circle_centre.dot(position_t_right.head<2>());
+
+      return {wf.angle_to_weight.inside_from_cos(dot),
+              wf.angle_to_weight.outside_from_cos(dot)};
+    }
+  }
+
+  std::tuple<bool, bool> WeightingFunctionEdges::point_inside_outside(
+      const Eigen::Ref<const Eigen::Vector3d> &position) const {
+    return point_inside_outside_t(wf._flippedBasis * position);
+  }
+
+  std::tuple<bool, bool> WeightingFunctionEdges::spherical_tri_inside_outside(
+      const Eigen::Ref<Eigen::Vector3d> &p1,
+      const Eigen::Ref<Eigen::Vector3d> &p2,
+      const Eigen::Ref<Eigen::Vector3d> &p3) const {
+    Eigen::Vector3d p1_t = wf._flippedBasis * p1;
+    Eigen::Vector3d p2_t = wf._flippedBasis * p2;
+    Eigen::Vector3d p3_t = wf._flippedBasis * p3;
+
+    bool p1_in, p1_out, p2_in, p2_out, p3_in, p3_out;
+
+    // find if the three corners are inside or outside; bail out early if they
+    // are inconsistent
+    std::tie(p1_in, p1_out) = point_inside_outside_t(p1_t);
+    std::tie(p2_in, p2_out) = point_inside_outside_t(p2_t);
+    if (p1_in != p2_in || p1_out != p2_out) return {false, false};
+    std::tie(p3_in, p3_out) = point_inside_outside_t(p3_t);
+    if (p2_in != p3_in || p2_out != p3_out) return {false, false};
+
+    if (p1_in && p2_in && p3_in) {
+      for (auto &p : points_outside)
+        if (pointInsideSphericalTriangle(p1_t, p2_t, p3_t, p))
+          return {false, false};
+
+      for (auto &edge : inside_edges) {
+        if (edge.intersects(p1_t, p2_t) || edge.intersects(p2_t, p3_t) ||
+            edge.intersects(p3_t, p1_t))
+          return {false, false};
+      }
+
+      return {true, false};
+    } else if (p1_out && p2_out && p3_out) {
+      for (auto &p : points_inside)
+        if (pointInsideSphericalTriangle(p1_t, p2_t, p3_t, p))
+          return {false, false};
+
+      for (auto &edge : outside_edges) {
+        if (edge.intersects(p1_t, p2_t) || edge.intersects(p2_t, p3_t) ||
+            edge.intersects(p3_t, p1_t))
+          return {false, false};
+      }
+      return {false, true};
+    } else
+      return {false, false};
+  }
+
+  std::tuple<bool, bool> WeightingFunctionEdges::spherical_tri_inside_outside(
```

Comment:
could this return a struct with sensibly named bools (inside, outside?) instead of a tuple?

Reply:
usually makes things easier to read and I've a vague memory of std::tuple doing some fairly complicated things that can slow down compile time if nothing else.